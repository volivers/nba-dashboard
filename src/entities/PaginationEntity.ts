import { Pagination } from '../interfaces/Pagination';

export interface PaginationEntity {
  current_page: number;
  next_page: number;
  per_page: number;
}

export const deserializePaginationEntity = (entity: PaginationEntity): Pagination => ({
  currentPage: entity.current_page,
  nextPage: entity.next_page,
  perPage: entity.per_page,
});

export const mockedPaginationEntity: PaginationEntity = {
  current_page: 1,
  next_page: 2,
  per_page: 25,
};
