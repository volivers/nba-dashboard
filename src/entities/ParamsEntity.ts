export interface ParamsEntity {
  page?: number;
  per_page?: number;
  game_ids?: number[];
  player_ids?: number[];
  team_ids?: number[];
}
