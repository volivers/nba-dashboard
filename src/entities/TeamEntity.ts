import { Team } from '../interfaces/Team';

export interface TeamEntity {
  abbreviation: string;
  city: string;
  conference: string;
  division: string;
  full_name: string;
  id: number;
  name: string;
}

export const deserializeTeamEntity = (entity: TeamEntity): Team => ({
  abbreviation: entity?.abbreviation,
  city: entity?.city,
  conference: entity?.conference,
  division: entity?.division,
  fullName: entity?.full_name,
  id: entity?.id,
  name: entity?.name,
});

export const mockedHomeTeamEntity: TeamEntity = {
  abbreviation: 'BOS',
  city: 'Boston',
  conference: 'Conference',
  division: 'Division',
  full_name: 'Boston Celtics',
  id: 2,
  name: 'Celtics',
};

export const mockedAwayTeamEntity: TeamEntity = {
  abbreviation: 'LAC',
  city: 'Los Angeles',
  conference: 'Conference',
  division: 'Division',
  full_name: 'LA Clippers',
  id: 1,
  name: 'Clippers',
};