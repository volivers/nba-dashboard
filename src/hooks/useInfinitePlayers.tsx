import { useInfiniteQuery } from 'react-query';
import { PlayerService } from '../services/PlayerService';

export const useInfinitePlayers = (teamId?: number) => useInfiniteQuery(
  ['players', teamId],
  ({ pageParam = 1 }) => PlayerService.getAll({ page: pageParam, perPage: 100 }),
  {
    getNextPageParam: (lastPage) => {
      if (!lastPage.meta.nextPage) return undefined;
      return lastPage.meta.currentPage + 1;
    },
  },
);
