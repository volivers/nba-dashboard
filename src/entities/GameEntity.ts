import { Game } from '../interfaces/Game';
import { TeamEntity, deserializeTeamEntity, mockedAwayTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface GameEntity {
  date: string;
  home_team: TeamEntity;
  home_team_score: number;
  id: number;
  period: number;
  postseason: boolean;
  season: number;
  status: string;
  time: string;
  visitor_team: TeamEntity;
  visitor_team_score: number;
}

export const deserializeGameEntity = (entity: GameEntity): Game => ({
  date: entity.date,
  homeTeam: deserializeTeamEntity(entity.home_team),
  homeTeamScore: entity.home_team_score,
  id: entity.id,
  period: entity.period,
  postseason: entity.postseason,
  season: entity.season,
  status: entity.status,
  time: entity.time,
  visitorTeam: deserializeTeamEntity(entity.visitor_team),
  visitorTeamScore: entity.visitor_team_score,
});

export const mockedGameEntity: GameEntity = {
  date: '2000-11-01T00:00:00.000Z',
  home_team: mockedHomeTeamEntity,
  home_team_score: 100,
  id: 1,
  period: 4,
  postseason: true,
  season: 1,
  status: 'Final',
  time: '',
  visitor_team: mockedAwayTeamEntity,
  visitor_team_score: 80,
};
