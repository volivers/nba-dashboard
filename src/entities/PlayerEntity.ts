import { Player } from '../interfaces/Player';
import { TeamEntity, deserializeTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface PlayerEntity {
  first_name: string;
  height_feet?: number;
  height_inches?: number;
  id: number;
  last_name: string;
  position: string;
  team: TeamEntity;
  weight_pounds?: number;
}

export const deserializePlayerEntity = (entity: PlayerEntity): Player => ({
  firstName: entity.first_name,
  heightFeet: entity?.height_feet,
  heightInches: entity?.height_inches,
  id: entity.id,
  lastName: entity.last_name,
  position: entity.position,
  team: deserializeTeamEntity(entity.team),
  weightPounds: entity?.weight_pounds,
});

export const mockedPlayerEntity: PlayerEntity = {
  first_name: 'Neemias',
  id: 88,
  last_name: 'Queta',
  position: 'C',
  team: mockedHomeTeamEntity,
};
