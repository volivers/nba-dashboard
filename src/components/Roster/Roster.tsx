import React, { useMemo } from 'react';
import { useInfinitePlayers } from '../../hooks/useInfinitePlayers';
import { Dropdown } from 'react-bootstrap';
import InfiniteScroll from 'react-infinite-scroller';
import { Player } from '../../interfaces/Player';

export interface RosterProps {
  activeTeamId: number;
  onSelectPlayer: (player: Player) => void;
  selectedPlayer?: Player;
}

const Roster = ({ activeTeamId, onSelectPlayer, selectedPlayer }: RosterProps) => {
  const { data, fetchNextPage, hasNextPage, isFetchingNextPage } = useInfinitePlayers(activeTeamId);
  const { firstName, lastName } = selectedPlayer || {};

  const teamPlayers = useMemo(() => {
    return data?.pages
      .flatMap(page => page.data)
      .filter(({ team }) => team.id === activeTeamId);
  }, [data, activeTeamId]);

  return (
    <div>
      <h1 className="my-3">Team Players</h1>
      <p className="mb-4"><small>Pick a player to view the stats.</small></p>
      <Dropdown className="mb-5">
        <Dropdown.Toggle variant="dark">
          {selectedPlayer ? `${firstName} ${lastName}` : 'Select player'}
        </Dropdown.Toggle>
        <Dropdown.Menu style={{ maxHeight: '300px', overflowY: 'auto' }}>
          <InfiniteScroll
            pageStart={0}
            loadMore={() => fetchNextPage()}
            hasMore={hasNextPage || isFetchingNextPage}
            useWindow={false}
          >
            {teamPlayers?.map((player) => (
              <Dropdown.Item key={player.id} onClick={() => onSelectPlayer(player)}>
                {player.firstName} {player.lastName}
              </Dropdown.Item>
            ))}
            {isFetchingNextPage && <Dropdown.Item>Loading more...</Dropdown.Item>}
          </InfiniteScroll>
        </Dropdown.Menu>
      </Dropdown>
    </div>
  );
};

export default Roster;
