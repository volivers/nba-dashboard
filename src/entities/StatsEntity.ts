import { Stats } from '../interfaces/Stats';
import { GameEntity, deserializeGameEntity, mockedGameEntity } from './GameEntity';
import { PlayerEntity, deserializePlayerEntity, mockedPlayerEntity } from './PlayerEntity';
import { TeamEntity, deserializeTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface StatsEntity {
  ast: number;
  blk: number;
  dreb: number;
  fg3_pct: number;
  fg3a: number;
  fg3m: number;
  fg_pct: number;
  fga: number;
  fgm: number;
  ft_pct: number;
  fta: number;
  ftm: number;
  game: GameEntity;
  id: number;
  min: string;
  oreb: number;
  pf: number;
  player: PlayerEntity;
  pts: number;
  reb: number;
  stl: number;
  team: TeamEntity;
  turnover: number;
}

export const deserializeStatsEntity = (entity: StatsEntity): Stats => ({
  ast: entity.ast,
  blk: entity.blk,
  dreb: entity.dreb,
  fg3Pct: entity.fg3_pct,
  fg3a: entity.fg3a,
  fg3m: entity.fg3m,
  fgPct: entity.fg_pct,
  fga: entity.fga,
  fgm: entity.fgm,
  ftPct: entity.ft_pct,
  fta: entity.fta,
  ftm: entity.ftm,
  game: deserializeGameEntity(entity.game),
  id: entity.id,
  min: entity.min,
  oreb: entity.oreb,
  pf: entity.pf,
  player: deserializePlayerEntity(entity.player),
  pts: entity.pts,
  reb: entity.reb,
  stl: entity.stl,
  team: deserializeTeamEntity(entity.team),
  turnover: entity.turnover,
});

export const mockedStatsEntity: StatsEntity = {
  ast: 10,
  blk: 4,
  dreb: 15,
  fg3_pct: 1,
  fg3a: 10,
  fg3m: 10,
  fg_pct: 10,
  fga: 10,
  fgm: 10,
  ft_pct: 10,
  fta: 10,
  ftm: 10,
  game: mockedGameEntity,
  id: 1,
  min: '18',
  oreb: 5,
  pf: 14,
  player: mockedPlayerEntity,
  pts: 20,
  reb: 10,
  stl: 5,
  team: mockedHomeTeamEntity,
  turnover: 10,
};
