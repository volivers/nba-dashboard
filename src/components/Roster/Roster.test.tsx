import { render, screen } from '@testing-library/react';
import Roster, { RosterProps } from './Roster';
import { deserializePaginationEntity, mockedPaginationEntity } from '../../entities/PaginationEntity';
import { useInfinitePlayers } from '../../hooks/useInfinitePlayers';
import { deserializePlayerEntity, mockedPlayerEntity } from '../../entities/PlayerEntity';

jest.mock('../../hooks/useInfinitePlayers');
const mockedUseInfinitePlayers = useInfinitePlayers as jest.Mock;

describe('Roster', () => {
  let props: RosterProps;

  beforeEach(() => {
    props = {
      activeTeamId: 2,
      onSelectPlayer: jest.fn(),
    };

    mockedUseInfinitePlayers.mockResolvedValue({
      fetchNextPage: jest.fn(),
      hasNextPage: false,
      isFetchingNextPage: false,
      pages: [
        {
          data: [deserializePlayerEntity(mockedPlayerEntity)],
          meta: deserializePaginationEntity(mockedPaginationEntity),
        }
      ],
    });
  });

  it('renders properly', () => {
    const { container } = render(<Roster {...props} />);

    expect(container).toMatchSnapshot();
  });

  describe('when there is a selected player', () => {
    beforeEach(() => {
      props = {
        ...props,
        selectedPlayer: deserializePlayerEntity(mockedPlayerEntity),
      };
    });

    it('renders the player name', () => {
      render(<Roster {...props} />);

      expect(screen.getByText(/Neemias Queta/)).toBeInTheDocument();
    });
  });
});
