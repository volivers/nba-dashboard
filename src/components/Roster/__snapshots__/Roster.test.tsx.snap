// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roster renders properly 1`] = `
<div>
  <div>
    <h1
      class="my-3"
    >
      Team Players
    </h1>
    <p
      class="mb-4"
    >
      <small>
        Pick a player to view the stats.
      </small>
    </p>
    <div
      class="mb-5 dropdown"
    >
      <button
        aria-expanded="false"
        class="dropdown-toggle btn btn-dark"
        id="react-aria-:r0:"
        type="button"
      >
        Select player
      </button>
    </div>
  </div>
</div>
`;
